Rodhern's update 3.

Rodhern opened the pull request:
A possible merge as per #4.
I do not use KSP 1.5.1 myself, so I can't even know if this will compile.
The changes are:
 * Revised interpretations of some stability derivatives, such as the definitions of the 'right' and 'up' components of 'velocity', and inverted signs for Zq, Ze, Xq and Xe to better match the sign convention used for Zu and Xu.
 * Tweaked stability derivatives solver for level flight AoA.
 * Reset editor part cache at editor reset to fix the most frequent cause of the stability derivatives changing on craft reload (ferram4#177).
 * Minor updates to localization files.
 * Cosmetic changes to the stability derivatives GUI window.
 * Added stability derivatives export button(s) ([screenshots available](https://github.com/Rodhern/Ferram-Aerospace-Research/wiki)).

dkavolis commented on .gitignore:
```diff
@@ -4,6 +4,7 @@ obj
 *.mdb
 /GameData/FerramAerospaceResearch/Custom*.cfg
 /GameData/FerramAerospaceResearch/FARForceDataUpdate.cfg
+/note.txt
```
No need to include this in PR

Rodhern replied:
Just need a few minutes to figure out how to do this :-)

Rodhern replied:
I just pushed the 'removed' note.txt to the source branch. I hope this is the Github way that I am supposed to solve the comment.

dkavolis commented on FerramAerospaceResearch/FARGUI/FAREditorGUI/Simulation/InstantConditionSim.cs:
```diff
@@ -141,15 +141,15 @@ public void GetClCdCmSteady(InstantConditionSimInput input, out InstantCondition
-            Vector3d sideways = Vector3.Cross(velocity, liftVector).normalized;
+            Vector3d sideways = -forward * cosAlpha * sinBeta;
+            sideways += right * (cosPhi * cosBeta - sinPhi * sinAlpha * sinBeta); // Rodhern: The same as Vector3.Cross(velocity, liftVector)
+            sideways += up * (cosPhi * sinAlpha * sinBeta + sinPhi * cosBeta);    //  but possibly with better accuracy.
```
It's better to leave it as it was because Vector3.cross is more concise and better for readability unless the accuracy is much worse

Rodhern replied:
Sure I can change that. Do you feel it is better for readability even when the remark is in the code? I feel there is sort of a nice analogous feel between 'velocity', 'liftVector' and 'sideways' this way. It is up to you though.

dkavolis replied:
We're computing principal axes so cross is easier and shows that explicitly, no one will ever bother computing the last component from scratch (at least on paper) when it is also perpendicular to the other two

dkavolis commented on FerramAerospaceResearch/FARGUI/FAREditorGUI/Simulation/InstantConditionSim.cs:
```diff
@@ -141,15 +141,15 @@ public void GetClCdCmSteady(InstantConditionSimInput input, out InstantCondition
-
-            velocity.Normalize();
+            velocity += right * (sinPhi * sinAlpha * cosBeta + cosPhi * sinBeta); // Rodhern: Changed velocity 'right' and 'up' components
+            velocity += -up * (cosPhi * sinAlpha * cosBeta - sinPhi * sinBeta);   //  to align vector to naive forward direction.
```
Which order are the rotations applied in? Need to be sure the math is correct before merging

Rodhern replied:
Are you sure you used the formulas from my code? Did you change the components back to the code from before the merge on purpose?

dkavolis commented on FerramAerospaceResearch/FARGUI/FAREditorGUI/Simulation/StabilityDerivCalculator.cs:
```diff
@@ -292,8 +316,8 @@ public StabilityDerivOutput CalculateStabilityDerivs(double u0, double q, double
-            pertOutput.Cl *= q * area / mass;
-            pertOutput.Cd *= q * area / mass;
+            pertOutput.Cl *= -q * area / mass; // Rodhern: Replaced 'q' by '-q', so that formulas
+            pertOutput.Cd *= -q * area / mass; //  for Ze and Xe match those for Zu and Xu.
```
I'd be careful about changing signs, I'm sure ferram had a good reason to set them the way they were. If anything, the negative signs shouldn't be there, the equations follow https://courses.cit.cornell.edu/mae5070/DynamicEquations.pdf (for the most part)

Rodhern replied:
You can see my reason for changing signs here: [Kerbal Space Program Forum](https://forum.kerbalspaceprogram.com/index.php?/topic/109098-official-far-craft-repository/&do=findComment&comment=2425057). It would be great if you can point out where it is that my reasoning fails.

dkavolis replied:
You're probably right, at the very least the signs should be consistent for the same axes. I'm not sure about the need for https://github.com/dkavolis/Ferram-Aerospace-Research/blob/d9665fba79aa90b32226a5e60a95c3540b35109b/FerramAerospaceResearch/FARGUI/FAREditorGUI/Simulation/StabilityDerivCalculator.cs#L235-L236
according to the equations.

Rodhern replied:
I have not really paid attention to that. Let us run through some quick back of the envelope calculations and verify that it makes sense to us. Let us look at Zw (i.e. dZ/dw).

The initial state is the aircraft flying horizontally at speed u0. In fact moving at speed u = u0 (forward), w = 0 (downward) and v = 0 (rightward). The angle of attack is some alpha.

Now imagine that we manipulate the direction of travel by rotating it ever so slightly towards the ground. We are in effect increasing alpha, and this will cause two changes that we will monitor. It changes the velocity vector and it changes the downward force. Say the rotation amount is tau radians. As an example of a specific near zero numerical value of tau I will use the symbol 2°.

The velocity deltas are,
 dw = sin(tau) * u0 - 0 ≈ 2° * u0,
 du = cos(tau) * u0 - u0 ≈ 0,
 dv = 0.

The downward force delta (or in our case actually the acceleration delta) is, for some relevant constant k,
 dZ = -cos(tau) * k * u0² * CL(AoA= alpha + tau)  // the new lift component
    + -sin(tau) * k * u0² * CD(AoA= alpha + tau)  // the new drag component
    - -cos(  0) * k * u0² * CL(AoA= alpha +   0)  // the old downward force
  ≈ tau * d( -cos(tau) * k * u0² * CL(tau) + -sin(tau) * k * u0² * CD(tau) )
  = tau * k * u0² * ((-cos * dCL) + (sin * CL) + (-sin * dCD) + (-cos * CD))
  ≈ 2° * k * u0² * (-dCL + -CD)

You probably guessed by now that CL(tau) is the coefficient of lift in the given circumstances at an angle of attack equal to alpha plus tau, that CD(tau) is the corresponding coefficient of drag, and that dCL and dCD are the (mathematical) derivatives of CL and CD with respect to the angle of attack parameter (in radians).

Now if q = 1/2 * density * u0² and k = 1/2 * density * area / mass we get
 Zw = dZ/dw ≈ 2° * k * u0² * (-dCL + -CD) / (2° * u0)
  = -k * u0² * (dCL + CD) / u0
  = -(1/2 * density * area / mass) * u0² * (dCL + CD) / u0
  = -q * area * (dCL + CD) / (mass * u0)

Now compare to the code:
 dCL: pertOutput.Cl = (pertOutput.Cl - nominalOutput.Cl) / (2 * FARMathUtil.deg2rad);
 (dCL + CD): pertOutput.Cl += nominalOutput.Cd;
 -q * area * (dCL + CD) / (mass * u0): pertOutput.Cl *= -q * area / (mass * u0);

I think it makes sense. Please point out mistakes that I make so that we do not end up confusing other comment readers.

dkavolis commented on FerramAerospaceResearch/FARGUI/FAREditorGUI/Simulation/InstantConditionSim.cs:
```diff
@@ -158,7 +158,7 @@ public void GetClCdCmSteady(InstantConditionSimInput input, out InstantCondition
                     continue;
 
-                w.ComputeForceEditor(velocity.normalized, input.machNumber, 2);
+                w.ComputeForceEditor(velocity, input.machNumber, 2);
```
`velocity` is no longer normalized here, this will affect force computations

Rodhern replied:
This question is not meant in a harsh way; I am just a bit clumsy writing it succinct.
Can you produce an example where the velocity.normalized is (strictly) closer to normalized than the current ('not normalized') velocity?
If so that may also be a good reason to use Vector3.cross for sideways, that you mentioned in an earlier comment.

dkavolis replied:
You're right, sorry. Had a mistake in my calculations. Still, the floating point errors can accumulate so I would normalize the vector afterwards anyway.

Rodhern replied:
Sure. We can do that to make it more easy to read the intention of the formulas.
(for what it is worth I suspect that the direct formulas uses double precision floating point math and the Normalize have to deal with single precision and won't improve anything).

dkavolis commented on GameData/FerramAerospaceResearch/FAR.version:
```diff
@@ -10,7 +10,7 @@
     "MINOR" : 15,
     "PATCH" : 9,
-    "BUILD" : 5
+    "BUILD" : 6
```
No need to change the version, I will do it when I publish a new release

Rodhern replied:
Roger, will set it back to 5.

dkavolis commented:
Thanks for your PR. I just have a few questions and suggestions.

* You replaced some spaces in localization files with explicit unicode, any reason a simple space is not sufficient? You have also removed some spaces such that values following them will appear without one.
* Instead of `Loop` in export it might be better to use `Sweep`.
* The input values of altitude and Mach number are taken from a file deep in KSP folder, I think it would be better to have a separate tab open where the derivatives are for inputting the values manually (which could be saved in config.xml). I was thinking of having an input list for each, something like in Paraview:

    ![image](https://user-images.githubusercontent.com/12998363/48317339-41911280-e5e8-11e8-9b06-b842f01e9c75.png)

    where the user could input values one by one or as a lin/logscale.
* The export format is inefficient (outputs too much data in terms of definitions) and will probably be difficult to read with existing tools (I'm thinking Python). What do you think about outputting the results in csv format? Body name, aircraft properties and inertia could be left in the header and the rest in a table format. It would make importing the file that much easier into existing tools.
* The exported files are stored in FerramAerospaceResearh/Plugins/PluginData which can be confusing. It might be better to store all the files in FerramAerospaceResearch/StabilityDerivatives or something similar, the default file names should also be changed so they don't overwrite each other, maybe "{CraftName}_{datetime}" such that they are unique.
* Running the export at multiple points freezes the game for a short time, this should be offloaded to a different thread.
* What are the benefits of your `SillySearchMethod` over `BrentsMethod`? The latter is robust, fast and popular root finding algorithm.

Rodhern commented:
>     * You replaced some spaces in localization files with explicit unicode, any reason a simple space is not sufficient? You have also removed some spaces such that values following them will appear without one.
In my KSP 1.3.1 version only the explicit unicode spaces show up in the GUI. That is the reason they are made explicit. The ones that are removed are to show clearer that the characters will not appear.

>     * Instead of `Loop` in export it might be better to use `Sweep`.
That makes sense; thanks.

>     * The input values of altitude and Mach number are taken from a file deep in KSP folder, I think it would be better to have a separate tab open where the derivatives are for inputting the values manually (which could be saved in config.xml). I was thinking of having an input list for each, something like in Paraview:
>       ![image](https://user-images.githubusercontent.com/12998363/48317339-41911280-e5e8-11e8-9b06-b842f01e9c75.png)
>       where the user could input values one by one or as a lin/logscale.
Sounds good to me, but I am not able to code that myself.

>     * The export format is inefficient (outputs too much data in terms of definitions) and will probably be difficult to read with existing tools (I'm thinking Python). What do you think about outputting the results in csv format? Body name, aircraft properties and inertia could be left in the header and the rest in a table format. It would make importing the file that much easier into existing tools.
Sound good too. Again I am not sure I have the skill and/or time to make it so. Feel free to take what can be reused and create some better export functionality.

>     * The exported files are stored in FerramAerospaceResearh/Plugins/PluginData which can be confusing. It might be better to store all the files in FerramAerospaceResearch/StabilityDerivatives or something similar, the default file names should also be changed so they don't overwrite each other, maybe "{CraftName}_{datetime}" such that they are unique.
:-) 

>     * Running the export at multiple points freezes the game for a short time, this should be offloaded to a different thread.
:-)

>     * What are the benefits of your `SillySearchMethod` over `BrentsMethod`? The latter is robust, fast and popular root finding algorithm.
The benefit is that BrentsMethod won't (as easily) find the maximum CL, which in turns produces an inability to solve the AoA for slow flight. The SillySearchMethod did better in my examples.

dkavolis commented:
> Sounds good to me, but I am not able to code that myself.

I will probably give it a try then when I have more time.

> Sound good too. Again I am not sure I have the skill and/or time to make it so. Feel free to take what can be reused and create some better export functionality.

This shouldn't be too difficult to implement, once everything is done a contour plot might be a nice additional feature.

> The benefit is that BrentsMethod won't (as easily) find the maximum CL, which in turns produces an inability to solve the AoA for slow flight. The SillySearchMethod did better in my examples.

Then a switch based on Mach number could be used.

Rodhern commented:
> This shouldn't be too difficult to implement, once everything is done a contour plot might be a nice additional feature.
Agreed, that would be pretty cool.

> Then a switch based on Mach number could be used.
Excellent idea. So it would kind of go: First use Brent, if that fails then go Silly ?

dkavolis commented:
> Excellent idea. So it would kind of go: First use Brent, if that fails then go Silly ?

Simply if mach number is below, say, 0.3, use `SillySearchMethod`, otherwise use `BrentsMethod`. The best switch value should be slightly larger than the lowest `BrentsMethod` fails on.

Rodhern commented:
> Simply if mach number is below, say, 0.3, use `SillySearchMethod`, otherwise use `BrentsMethod`. The best switch value should be slightly larger than the lowest `BrentsMethod` fails on.

Sounds fine. My only concern was that the threshold speed might be quite craft dependent.

dkavolis commented:
You can then add a check for failure but a fixed Mach switch should be faster on average.

Rodhern commented:
> You can then add a check for failure but a fixed Mach switch should be faster on average.

Sounds good. I will see when I get time to look at the code again, and get it fixed, maybe next weekend, maybe later. If you feel like tinkering be my guest and go ahead before then.

Rodhern commented:
> Do you mind submitting separate PRs for each of the changes? It would make keeping track of changes easier in case ferram ever returns and some of them can be merged already.

I guess I can do that. I will plan to probably do that some time.
(I am already a bit surprised at how Github does PRs; that the branch tips just keep chugging along).

Rodhern commented:
> Do you mind submitting the changes without sweep/export of the stability derivatives (can be a single PR) so I can merge them in?

That sounds like a plan to me. I will try to find some time in the new year to make an updated pull request to one of your newer branches.

I will do it as a single new untested (because I do not play any of the newest KSP versions and cannot check to see if the merged source compiles) PR, once I have removed / commented out the stability derivatives export buttons in the GUI, and added (some or most of) the code readability enhancing normalizations and tweaks discussed in the comments above.

Rodhern commented:
Closing this PR; it is to be replaced by a newer PR.
